File: scripts/montar-pieza-porta-completos-bulldog.py
#!/usr/bin/env python3
"""
Monta la pieza 1080×1920 con la FOTO REAL del porta completos bulldog
(sin regenerar el producto con IA).

Uso (PC):
  1) Guardá la foto del producto como:
       index\\clientes\\impresoreando\\piezas\\foto-producto-bulldog.jpg
     (también busca en Descargas si no está)
  2) python scripts/montar-pieza-porta-completos-bulldog.py

Salida:
  index/clientes/impresoreando/piezas/porta-completos-bulldog-1080x1920.png
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont, ImageFilter, ImageOps

ROOT = Path(__file__).resolve().parents[1]
PIEZAS = ROOT / "index" / "clientes" / "impresoreando" / "piezas"
OUT = PIEZAS / "porta-completos-bulldog-1080x1920.png"
W, H = 1080, 1920
BG = (232, 220, 200)  # beige identidad
INK = (62, 42, 28)  # café tipografía


def candidatos_foto() -> list[Path]:
    downloads = Path(os.environ.get("USERPROFILE") or os.environ.get("HOME") or "") / "Downloads"
    names = [
        "foto-producto-bulldog.jpg",
        "foto-producto-bulldog.jpeg",
        "foto-producto-bulldog.png",
        "porta-completos-bulldog.jpg",
        "porta-completos-bulldog.jpeg",
        "porta-completos-bulldog.png",
    ]
    paths: list[Path] = []
    for base in (PIEZAS, downloads):
        if not base.exists():
            continue
        for n in names:
            p = base / n
            if p.is_file():
                paths.append(p)
    return paths


def font(size: int, bold: bool = False) -> ImageFont.FreeTypeFont | ImageFont.ImageFont:
    candidates = [
        "C:/Windows/Fonts/georgia.ttf",
        "C:/Windows/Fonts/times.ttf",
        "C:/Windows/Fonts/timesbd.ttf" if bold else None,
        "/usr/share/fonts/truetype/dejavu/DejaVuSerif-Bold.ttf" if bold else "/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf",
        "/usr/share/fonts/truetype/liberation/LiberationSerif-Bold.ttf" if bold else "/usr/share/fonts/truetype/liberation/LiberationSerif-Regular.ttf",
        "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    ]
    for c in candidates:
        if c and Path(c).is_file():
            try:
                return ImageFont.truetype(c, size)
            except OSError:
                pass
    return ImageFont.load_default()


def text_size(draw: ImageDraw.ImageDraw, text: str, fnt) -> tuple[int, int]:
    box = draw.textbbox((0, 0), text, font=fnt)
    return box[2] - box[0], box[3] - box[1]


def main() -> int:
    fotos = candidatos_foto()
    if not fotos:
        print(
            "[error] No encontré la foto del producto.\n"
            "  Guardala como:\n"
            f"    {PIEZAS / 'foto-producto-bulldog.jpg'}\n"
            "  o en Descargas con el mismo nombre, y re-ejecutá."
        )
        return 1

    foto_path = fotos[0]
    print("[foto]", foto_path)
    product = Image.open(foto_path).convert("RGBA")

    # Recorte centrado suave: quitar bordes extremos si hay mucha pared/mano
    # (no inventa el producto: solo encuadra)
    pw, ph = product.size
    # Si la foto es muy vertical con mucha pared arriba, recortar un poco arriba/abajo
    top = int(ph * 0.02)
    bottom = int(ph * 0.98)
    left = int(pw * 0.02)
    right = int(pw * 0.98)
    product = product.crop((left, top, right, bottom))

    canvas = Image.new("RGB", (W, H), BG)
    draw = ImageDraw.Draw(canvas)

    title_f = font(92, bold=True)
    sub_f = font(36, bold=False)
    foot_f = font(28, bold=False)

    title = "Porta completos"
    sub = "El bulldog que cuida tu completo"
    foot = "Hecho a pedido"

    tw, th = text_size(draw, title, title_f)
    sw, sh = text_size(draw, sub, sub_f)
    fw, fh = text_size(draw, foot, foot_f)

    draw.text(((W - tw) / 2, 110), title, font=title_f, fill=INK)
    draw.text(((W - sw) / 2, 110 + th + 18), sub, font=sub_f, fill=INK)

    # Zona producto
    top_zone = 110 + th + 18 + sh + 50
    bottom_zone = H - 160
    max_w = int(W * 0.86)
    max_h = bottom_zone - top_zone
    product_fit = ImageOps.contain(product, (max_w, max_h), Image.Resampling.LANCZOS)

    # Sombra suave
    shadow = Image.new("RGBA", (W, H), (0, 0, 0, 0))
    sx = (W - product_fit.width) // 2 + 12
    sy = top_zone + (max_h - product_fit.height) // 2 + 18
    sh_img = Image.new("RGBA", product_fit.size, (0, 0, 0, 55))
    shadow.paste(sh_img, (sx, sy), sh_img)
    shadow = shadow.filter(ImageFilter.GaussianBlur(18))
    canvas = Image.alpha_composite(canvas.convert("RGBA"), shadow).convert("RGB")
    draw = ImageDraw.Draw(canvas)

    px = (W - product_fit.width) // 2
    py = top_zone + (max_h - product_fit.height) // 2
    if product_fit.mode == "RGBA":
        canvas.paste(product_fit, (px, py), product_fit)
    else:
        canvas.paste(product_fit.convert("RGB"), (px, py))

    # Footer
    fy = H - 100
    line_y = fy + fh // 2
    gap = 24
    lx1, lx2 = 80, (W - fw) // 2 - gap
    rx1, rx2 = (W + fw) // 2 + gap, W - 80
    draw.line((lx1, line_y, lx2, line_y), fill=INK, width=2)
    draw.line((rx1, line_y, rx2, line_y), fill=INK, width=2)
    draw.text(((W - fw) / 2, fy), foot, font=foot_f, fill=INK)

    PIEZAS.mkdir(parents=True, exist_ok=True)
    canvas.save(OUT, "PNG", optimize=True)
    # alias viejo
    alias = PIEZAS / "porta-celular-bulldog-1080x1920.png"
    canvas.save(alias, "PNG", optimize=True)
    print("[ok]", OUT, f"{W}x{H}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/montar-pieza-porta-completos-bulldog.js
#!/usr/bin/env node
/**
 * Monta pieza IG 1080×1920 Porta Completo Bulldog (PCPEBUL001).
 *
 * Preferencia: foto real del producto (la de Instagram).
 *   index/clientes/impresoreando/piezas/foto-producto-bulldog.jpg
 * Fallback: piezas/porta-completo-bulldog-producto-ref.png
 *
 *   node scripts/montar-pieza-porta-completos-bulldog.js
 *
 * Requiere: Python3 + Pillow (pip install pillow)
 */
const { spawnSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const ROOT = path.join(__dirname, '..');
const PIEZAS = path.join(ROOT, 'index/clientes/impresoreando/piezas');
const REDES = path.join(ROOT, 'index/clientes/impresoreando/redes');
const LOGO = path.join(ROOT, 'index/clientes/impresoreando/identidad/logo-ima2.png');
const OUT = path.join(PIEZAS, 'porta-completos-bulldog-1080x1920.png');
const OUT2 = path.join(REDES, 'porta-completos-bulldog-1080x1920.png');

const foto =
  [
    path.join(PIEZAS, 'foto-producto-bulldog.jpg'),
    path.join(PIEZAS, 'foto-producto-bulldog.jpeg'),
    path.join(PIEZAS, 'foto-producto-bulldog.png'),
    path.join(PIEZAS, 'porta-completo-bulldog-producto-ref.jpg'),
    path.join(PIEZAS, 'porta-completo-bulldog-producto-ref.png'),
  ].find((p) => fs.existsSync(p)) || null;

if (!foto) {
  console.error('Falta foto del producto. Poné la de Instagram en:');
  console.error('  index/clientes/impresoreando/piezas/foto-producto-bulldog.jpg');
  process.exit(1);
}

const py = `
from PIL import Image, ImageDraw, ImageFont
from pathlib import Path
W,H=1080,1920
NAVY=(22,58,107); ORANGE=(240,122,40); CREAM=(247,241,230); CREAM2=(238,230,214); WHITE=(255,253,247)
img=Image.new('RGB',(W,H),CREAM)
d=ImageDraw.Draw(img)
for y in range(H):
  t=y/(H-1)
  d.line([(0,y),(W,y)], fill=(int(CREAM[0]*(1-t)+CREAM2[0]*t), int(CREAM[1]*(1-t)+CREAM2[1]*t), int(CREAM[2]*(1-t)+CREAM2[2]*t)))
d.rectangle([0,0,W,12], fill=ORANGE)
logo=Image.open(r'''${LOGO.replace(/\\/g, '/')}''').convert('RGBA')
lw=560; lh=int(logo.height*(lw/logo.width)); logo=logo.resize((lw,lh), Image.Resampling.LANCZOS)
pad_x,pad_y=32,24; cw,ch=lw+pad_x*2, lh+pad_y*2
card=Image.new('RGBA',(cw,ch),(255,253,247,255)); cd=ImageDraw.Draw(card)
cd.rounded_rectangle([0,0,cw-1,ch-1], radius=26, outline=(22,58,107,90), width=3)
lx=(W-cw)//2; ly=56
img.paste(card,(lx,ly),card); img.paste(logo,(lx+pad_x,ly+pad_y),logo)
prod=Image.open(r'''${foto.replace(/\\/g, '/')}''').convert('RGB')
tw,th=920,1120
scale=max(tw/prod.width, th/prod.height)
prod=prod.resize((int(prod.width*scale), int(prod.height*scale)), Image.Resampling.LANCZOS)
prod=prod.crop(((prod.width-tw)//2,(prod.height-th)//2,(prod.width-tw)//2+tw,(prod.height-th)//2+th))
mask=Image.new('L',(tw,th),0); ImageDraw.Draw(mask).rounded_rectangle([0,0,tw-1,th-1], radius=40, fill=255)
frame=Image.new('RGB',(tw+10,th+10),WHITE); fd=ImageDraw.Draw(frame)
fd.rounded_rectangle([0,0,tw+9,th+9], radius=44, outline=NAVY, width=5)
fx=(W-frame.width)//2; fy=ly+ch+36
img.paste(frame,(fx,fy)); img.paste(prod,(fx+5,fy+5),mask)
def font(sz,bold=False):
  p='/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf' if bold else '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf'
  return ImageFont.truetype(p,sz) if Path(p).exists() else ImageFont.load_default()
f_sku,f_title,f_meta,f_handle=font(34,True),font(48,True),font(30,True),font(28,False)
sku='PCPEBUL001'; bb=d.textbbox((0,0),sku,font=f_sku); pw,ph=bb[2]-bb[0]+48, bb[3]-bb[1]+26
px=(W-pw)//2; py=fy+frame.height+32
d.rounded_rectangle([px,py,px+pw,py+ph], radius=999, fill=NAVY)
d.text((px+24,py+10),sku,font=f_sku,fill=WHITE)
title='Porta Completo Perro Bulldog'; tb=d.textbbox((0,0),title,font=f_title)
tx=(W-(tb[2]-tb[0]))//2; ty=py+ph+22
d.text((tx,ty),title,font=f_title,fill=NAVY)
rb=d.textbbox((0,0),'@impresoreando',font=f_handle); my=ty+(tb[3]-tb[1])+28
d.text((72,my),'Todo es a pedido',font=f_meta,fill=ORANGE)
d.text((W-72-(rb[2]-rb[0]),my),'@impresoreando',font=f_handle,fill=NAVY)
d.rectangle([0,H-16,W,H], fill=NAVY); d.rectangle([0,H-16,240,H], fill=ORANGE)
Path(r'''${OUT.replace(/\\/g, '/')}''').parent.mkdir(parents=True, exist_ok=True)
Path(r'''${OUT2.replace(/\\/g, '/')}''').parent.mkdir(parents=True, exist_ok=True)
img.save(r'''${OUT.replace(/\\/g, '/')}''','PNG',optimize=True)
img.save(r'''${OUT2.replace(/\\/g, '/')}''','PNG',optimize=True)
print('OK', r'''${OUT.replace(/\\/g, '/')}''')
print('foto:', r'''${foto.replace(/\\/g, '/')}''')
`;

const r = spawnSync('python3', ['-c', py], { encoding: 'utf8' });
if (r.stdout) process.stdout.write(r.stdout);
if (r.stderr) process.stderr.write(r.stderr);
if (r.status) process.exit(r.status);
console.log('Ver: index/clientes/impresoreando/piezas/porta-completos-bulldog-1080x1920.png');
console.log('IG:  http://127.0.0.1:8000/index/clientes/impresoreando/piezas/porta-completos-bulldog-1080x1920.png');
